metadata_layer.py

```diff
@@ -352,93 +352,492 @@ def __repr__(self):
+                    self.formats.append(
+                        FormatData(
+                            content_type=link.media_type,
+                            drm_scheme=DeliveryMechanism.NO_DRM,
+                            link=link
+                        )
                 )
+
+
+    def __repr__(self):
+        description_string = '<CirculationData primary_identifier=%(primary_identifier)r| licenses_owned=%(licenses_owned)s|'
+        description_string += ' licenses_available=%(licenses_available)s| default_rights_uri=%(default_rights_uri)s|' 
+        description_string += ' links=%(links)r| formats=%(formats)r| data_source=%(data_source)s|>'
+
+        description_data = {'primary_identifier':self.primary_identifier, 'licenses_owned':self.licenses_owned}
+        description_data['licenses_available'] = self.licenses_available
+        description_data['default_rights_uri'] = self.default_rights_uri
+        description_data['links'] = self.links
+        description_data['formats'] = self.formats
+        description_data['data_source'] = self.data_source_name
+            
+        return description_string % description_data
+    
+
+    def data_source(self, _db):
+        if not self.data_source_obj:
+            if self._data_source:
+                obj = DataSource.lookup(_db, self._data_source)
+                if not obj:
+                    raise ValueError("Data source %s not found!" % self._data_source)
+                if not obj.offers_licenses:
+                    raise ValueError("Data source %s does not offer licenses and cannot be used as a LicenseData data source." % self._data_source)
+            else:
+                obj = None
+            self.data_source_obj = obj
+        return self.data_source_obj
+
+
+    def license_pool(self, _db):
+        """Find or create a LicensePool object for this CirculationData."""
+        if not self.primary_identifier:
+            raise ValueError(
+                "Cannot find license pool: CirculationData has no primary identifier."
+            )
+
+        license_pool = None
+        is_new = False
+        
+        identifier_obj, ignore = self.primary_identifier.load(_db)
+        data_source = self.data_source(_db)
+        license_pool = get_one(
+            _db, LicensePool, data_source=data_source,
+            identifier=identifier_obj
+        )
+
+        if not license_pool:
+            rights_status = get_one(
+                _db, RightsStatus, uri=self.default_rights_uri
             )
-            # TODO: Also put this in the log.
 
-        changed = (license_pool.licenses_owned != self.licenses_owned or
-                   license_pool.licenses_available != self.licenses_available or
-                   license_pool.patrons_in_hold_queue != self.patrons_in_hold_queue or
-                   license_pool.licenses_reserved != self.licenses_reserved)
+            last_checked = datetime.datetime.utcnow()
```

Comment:
we need to support taking last_checked as an argument to circulation data for the 3m case